packages/palette_generator/lib/palette_generator.dart

```diff
@@ -244,40 +238,46 @@ class PaletteGenerator with Diagnosticable {
     }
   }
 
-  /// Returns a vibrant color from the palette. Might be null if an appropriate
-  /// target color could not be found.
-  PaletteColor get vibrantColor => selectedSwatches[PaletteTarget.vibrant];
+  /// Returns a vibrant color from the palette. If an appropriate target color
+  /// could not be found then properties `color` is set to grey color and
+  /// `population` to 1.
+  PaletteColor get vibrantColor => selectedSwatches[PaletteTarget.vibrant]!;
 
-  /// Returns a light and vibrant color from the palette. Might be null if an
-  /// appropriate target color could not be found.
+  /// Returns a light and vibrant color from the palette. If an appropriate
+  /// target color could not be found then properties `color` is set to
+  /// grey color and `population` to 1.
   PaletteColor get lightVibrantColor =>
-      selectedSwatches[PaletteTarget.lightVibrant];
+      selectedSwatches[PaletteTarget.lightVibrant]!;
 
-  /// Returns a dark and vibrant color from the palette. Might be null if an
-  /// appropriate target color could not be found.
+  /// Returns a dark and vibrant color from the palette. If an appropriate
+  /// target color could not be found then properties `color` is set to
+  /// grey color and `population` to 1.
   PaletteColor get darkVibrantColor =>
-      selectedSwatches[PaletteTarget.darkVibrant];
+      selectedSwatches[PaletteTarget.darkVibrant]!;
 
-  /// Returns a muted color from the palette. Might be null if an appropriate
-  /// target color could not be found.
-  PaletteColor get mutedColor => selectedSwatches[PaletteTarget.muted];
+  /// Returns a muted color from the palette. If an appropriate target
+  /// color could not be found then properties `color` is set to grey color
+  /// and `population` to 1.
+  PaletteColor get mutedColor => selectedSwatches[PaletteTarget.muted]!;
 
-  /// Returns a muted and light color from the palette. Might be null if an
-  /// appropriate target color could not be found.
+  /// Returns a muted and light color from the palette. If an appropriate
+  /// target color could not be found then properties `color` is set to
+  /// grey color and `population` to 1.
   PaletteColor get lightMutedColor =>
-      selectedSwatches[PaletteTarget.lightMuted];
+      selectedSwatches[PaletteTarget.lightMuted]!;
 
-  /// Returns a muted and dark color from the palette. Might be null if an
-  /// appropriate target color could not be found.
-  PaletteColor get darkMutedColor => selectedSwatches[PaletteTarget.darkMuted];
+  /// Returns a muted and dark color from the palette. If an appropriate
+  /// target color could not be found then properties `color` is set to
+  /// grey color and `population` to 1.
+  PaletteColor get darkMutedColor => selectedSwatches[PaletteTarget.darkMuted]!;
 
   /// The dominant color (the color with the largest population).
   PaletteColor get dominantColor => _dominantColor;
-  PaletteColor _dominantColor;
+  late PaletteColor _dominantColor;
```

Comment:
@gspencergoog Was _dominantColor intended to be nullable? It doesn't say so, and there's code that uses it unconditionally, but also code just below that was setting it to null in one case. I looked into the logic more and it wasn't obvious to me where, if you go down the generation chain that leads to that point, things should be asserted.

E.g., can we assert somewhere that the image has non-zero size and propagate non-null-ness through all the scoring, etc., or do all those need to be nullable for edge case imagse?

Reply:
It was meant to be nullable, because if there are no `paletteColors`, then there can't be a dominant color. But if the image is non-zero-sized, then there have to be some paletteColors.

I suppose we could assert that the image was non-zero size, and a lot of the issues would go away.  I wanted the image loading to be robust to image providers that return zero sized images, but I guess it's pretty degenerate if it does that.

Reply:
Hm, I guess with a provide it's potentially hard for a client to make guarantees about at compile time, so we probably don't want to assert. In which case we're stuck with most of the nullability.